### DISCUSSION TITLE
Replicate CLI workflow via Python bindings

### DISCUSSION THREAD

**Discussion by Enchufa2** [rpm-software-management/dnf5#1617] [Q&A]
With dnf-3, this is the minimal workflow to install a package matching the CLI output:

```python3
import sys
import dnf
import dnf.cli.progress
import dnf.cli.output

base = dnf.Base()
progress = dnf.cli.progress.MultiFileProgressMeter(fo=sys.stdout)

base.conf.read()
base.read_all_repos()
base.repos.all().set_progress_bar(progress)
base.update_cache()
base.fill_sack()

base.install("vim")
base.resolve()
base.download_packages(base.transaction.install_set, progress)
base.do_transaction(dnf.cli.output.CliTransactionDisplay())
base.close()
```

Saving this as `test.py` and executing, I get:

```bash
$ python3 test.py 
Fedora rawhide openh264 (From Cisco) - x86_64                         2.8 kB/s | 1.4 kB     00:00    
Fedora - Rawhide - Developmental packages for the next Fedora release  15 MB/s |  20 MB     00:01    
(1/7): xxd-9.1.660-1.fc41.x86_64.rpm                                  194 kB/s |  36 kB     00:00    
(2/7): which-2.21-42.fc41.x86_64.rpm                                  191 kB/s |  42 kB     00:00    
(3/7): vim-enhanced-9.1.660-1.fc41.x86_64.rpm                         7.3 MB/s | 1.9 MB     00:00    
(4/7): libsodium-1.0.20-2.fc41.x86_64.rpm                             349 kB/s | 176 kB     00:00    
(5/7): vim-filesystem-9.1.660-1.fc41.noarch.rpm                       530 kB/s |  17 kB     00:00    
(6/7): gpm-libs-1.20.7-48.fc41.x86_64.rpm                             645 kB/s |  20 kB     00:00    
(7/7): vim-common-9.1.660-1.fc41.x86_64.rpm                           8.7 MB/s | 7.9 MB     00:00    
  Preparing        :                                                                              1/1 
  Installing       : xxd-2:9.1.660-1.fc41.x86_64                                                  1/7 
  Installing       : which-2.21-42.fc41.x86_64                                                    2/7 
  Installing       : vim-filesystem-2:9.1.660-1.fc41.noarch                                       3/7 
  Installing       : vim-common-2:9.1.660-1.fc41.x86_64                                           4/7 
  Installing       : libsodium-1.0.20-2.fc41.x86_64                                               5/7 
  Installing       : gpm-libs-1.20.7-48.fc41.x86_64                                               6/7 
  Installing       : vim-enhanced-2:9.1.660-1.fc41.x86_64                                         7/7 
  Running scriptlet: vim-enhanced-2:9.1.660-1.fc41.x86_64                                         7/7
```

as expected. Now, I tried to do the same with dnf5 following the current docs, and this is where I got so far:

```python3
import libdnf5

base = libdnf5.base.Base()
base.load_config()
base.setup()
repo_sack = base.get_repo_sack()
repo_sack.create_repos_from_system_configuration()
repo_sack.load_repos(libdnf5.repo.Repo.Type_AVAILABLE)

goal = libdnf5.base.Goal(base)
goal.add_install("vim")
transaction = goal.resolve()
transaction.download()
transaction.run()
```

The first question is how do I replicate the CLI output here, like before. Buf if I execute this, I get:

```bash
$ python3 test5.py 
$ vim
bash: vim: command not found
```

No output, as expected, but the installation didn't work. So the second question is what am I missing?

**Comment by kontura** [comment]
There is not really an easy way how to replicate the CLI output, especially from Python. Basically you have to use callbacks, simple example to get you started is in the [tutorial](https://dnf5.readthedocs.io/en/latest/tutorial/transaction.html).
Unfortunately the documentation for the callbacks is not generated yet so you have to look into the code to see what is available:
- for download callbacks: https://github.com/rpm-software-management/dnf5/blob/main/include/libdnf5/repo/download_callbacks.hpp
- for transaction callbacks: https://github.com/rpm-software-management/dnf5/blob/main/include/libdnf5/rpm/transaction_callbacks.hpp

Regarding the failing snippet I believe the problem is that you don't load the `SYSTEM` repo (which contains what packages are installed) therefore the dependencies cannot be correctly resolved. If you don't specify the `libdnf5.repo.Repo.Type_AVAILABLE` and use just `repo_sack.load_repos()` it will load both `SYSTEM` and `AVAILABLE` and it should work.

In general the functions have return values and there are ways how to see what went wrong though I understand it is harder without docs and again you would have to look into the cpp dnf5/libdnf5 code  to see how to do it.

**Reply by Enchufa2** [reply]
Thanks, yes, I tried to follow the docs, but there are still a lot of missing spots as you said. Particularly,

- I found that, similar to dnf-3's `MultiFileProgressMeter`, there is a `MultiProgressBar` in the `libdnf5-cli` package. I was hoping that this could be added somehow as a callback, but didn't find how. Is it usable?
- The other missing piece would be something similar to dnf-3's `CliTransactionDisplay`. I've seen that the C++ part does have a `libdnf5-cli/output` namespace, but this is not exported in the Python bindings, only the `progressbar` above. Would it be possible to export it too and use it to replicate the CLI's output?

I mean, having a flexible callback system is great. But I wouldn't like to reinvent the wheel when it comes to the transaction info reported to the user. If we could just reuse what the CLI shows, that would be great.

**Comment by Enchufa2** [comment]
On a side note, these are other differences I found with respect to dnf-3:

- dnf-3's `install`, `remove`, etc., raise an exception when a package name does not exist. dnf5's `add_install`, etc., however, always succeed and return nothing; then the transaction may fail. If I would like to ensure that the transaction can be run (i.e. add only existing packages and collect the ones that don't), what's the most efficient way of doing this? Maybe creating a `PackageQuery`, filtering by name and checking the length of the query before calling `add_install` maybe? Other?
- What if `repo_sack.load_repos()`, `goal.resolve()`, `transaction.download()` or `transaction.run()` fail? (Or other pieces I may be missing). Do they raise an exception that I can handle? Are there any docs demonstrating this?
